File: src/leetcode/problem432_all_one_data_structure/AllOne.js
function ListNode(data, count = 1) {
    this.count = count;
    this.data = data;
    this.prev = null;
    this.succ = null;
}

ListNode.prototype.insertAfter = function insertAfter(node) {
    this.succ.prev = node;
    node.succ = this.succ;
    this.succ = node;
    node.prev = this;
};

ListNode.prototype.remove = function remove() {
    this.prev.succ = this.succ;
    this.succ.prev = this.prev;
};

/**
 * Initialize your data structure here.
 */
function AllOne() {
    const header = new ListNode(new Set(), 0);
    const tailer = new ListNode(new Set(), Infinity);
    header.succ = tailer;
    tailer.prev = header;

    const map = new Map(); // count => node
    map.set(0, header);
    map.set(Infinity, tailer);
    const cmap = new Map(); // key => count
    /**
     * Inserts a new key <Key> with value 1. Or increments an existing key by 1.
     * @param {string} key
     * @return {void}
     */
    const inc = key => {
        let count = cmap.get(key) || 0;
        count += 1;
        cmap.set(key, count); // 修正计数
        const last = map.get(count - 1); // 找到之前的节点
        let node = map.get(count); // 得到新阶层
        if (!node) {
            // 新阶层可能不存在
            node = new ListNode(new Set(), count);
            last.insertAfter(node); // 插入到之前阶层的下一阶层
            map.set(count, node); // 将该阶层加入到 map
        }
        node.data.add(key); // 加入新阶层（阶层跃升完毕）

        if (1 < count) {
            // 从原有阶层剔除掉
            last.data.delete(key); // 从之前节点删除，阶层跃升了。

            if (last.data.size === 0) {
                // 原阶层是否需要删除
                map.delete(count - 1); // 删除映射表
                last.remove(); // 从列表删除
            }
        }
    };

    /**
     * Decrements an existing key by 1. If Key's value is 1, remove it from the data structure.
     * @param {string} key
     * @return {void}
     */
    const dec = key => {
        let count = cmap.get(key); // 根据映射表找到该key的计数值
        if (!count) return; // 不存在则直接退出
        const last = map.get(count); // 找到原有阶层
        count -= 1;

        if (count) {
            cmap.set(key, count); // 更新计数映射表
            let node = map.get(count); // 尝试获取新阶层
            if (!node) {
                // 新阶层可能不存在，需要创建
                node = new ListNode(new Set(), count);
                last.prev.insertAfter(node); // 新阶层插入到之前阶层前序阶层的后面
                map.set(count, node); // 将该阶层加入的映射表
            }
            node.data.add(key); // 加入新阶层（阶层滑落完毕）
        } else {
            // 可能要直接消失
            cmap.delete(key); // 从计数映射表移除
        }
        last.data.delete(key); // 从原有阶层移除
        if (last.data.size === 0) {
            // 原阶层是否需要删除
            map.delete(count + 1); // 删除映射表
            last.remove(); // 从列表删除
        }
    };
    /**
     * Returns one of the keys with maximal value.
     * @return {string}
     */
    const getMaxKey = () => Array.from(tailer.prev.data)[0] || "";

    /**
     * Returns one of the keys with Minimal value.
     * @return {string}
     */
    const getMinKey = () => Array.from(header.succ.data)[0] || "";

    return { inc, dec, getMaxKey, getMinKey };
}

File: src/leetcode/problem432_all_one_data_structure/AllOne.java
package leetcode.problem432_all_one_data_structure;

import java.util.*;

/**
 * @author zhangxu create on 2019/1/15
 * version 1.0
 */
public class AllOne {
    private Map<String,Node> map=new HashMap();
    private Node tail;
    private Node head;
    class Node{
        public Node(String key,int value){
            this.value=value;
            this.keys=new ArrayList();
            keys.add(key);
        }
        public List<String> keys;
        public int value;
        public Node pre;
        public Node next;
    }
    /** Initialize your data structure here. */
    public AllOne() {

    }

    /** Inserts a new key <Key> with value 1. Or increments an existing key by 1. */
    public void inc(String key) {
        if(map.containsKey(key)){
            Node node=map.get(key);
            map.remove(key);
            node.keys.remove(key);
            int newValue=node.value+1;
            Node currentNext=node.next;

            if(currentNext==null||currentNext.value!=newValue){
                Node newNode=new Node(key,newValue);
                insert(node,newNode);
                map.put(key,newNode);
            }
            else {
                currentNext.keys.add(key);
                map.put(key,node.next);
            }
            if(node.keys.size()==0){
                delete(node);
            }
        }
        else {
            if(head!=null&&head.value==1){
                head.keys.add(key);
                map.put(key,head);
            }
            else {
                Node newNode=new Node(key,1);
                insert(head,newNode);
                head=newNode;
                map.put(key,newNode);
            }

        }
    }

    /** Decrements an existing key by 1. If Key's value is 1, remove it from the data structure. */
    public void dec(String key) {
        if(map.containsKey(key)){
            Node node=map.get(key);
            node.keys.remove(key);
            map.remove(key);
            if(node.value!=1){
                int newValue=node.value-1;
                if(node.pre!=null&&node.pre.value==newValue){
                    node.pre.keys.add(key);
                    map.put(key,node.pre);
                }
                else{
                    Node newNode=new Node(key,newValue);
                    insert(node.pre,newNode);
                    map.put(key,newNode);
                }
            }
            if(node.keys.size()==0){
                delete(node);
            }
        }
    }

    /** Returns one of the keys with maximal value. */
    public String getMaxKey() {
        if(head==null){
            return "";
        }
        return head.keys.get(0);
    }

    /** Returns one of the keys with Minimal value. */
    public String getMinKey() {
        if(tail==null){
            return "";
        }
        return tail.keys.get(0);
    }
    public void insert(Node pre,Node now){
        if(head==null&&tail==null){
            head=now;
            tail=now;
        }
        else if(pre==head&&pre==tail){
            head.next=now;
            now.pre=head;
            tail=now;
        }
        else if(pre==head){
            Node headNext=pre.next;
            pre.next=now;
            now.pre=head;
            now.next=headNext;
            headNext.pre=now;
        }
        else if(pre==tail){
            tail.next=now;
            now.pre=tail;
            tail=now;
        }
        else {
            Node next=pre.next;
            pre.next=now;
            now.pre=pre;
            now.next=next;
            next.pre=now;
        }
    }
    public void delete(Node now){
        if(head==now&&tail==now){
            tail=null;
            head=null;
        }
        else if(head==now){
            head=head.next;
            head.pre=null;
        }
        else if(tail==null){
            tail=tail.pre;
            tail.next=null;
        }
        else {
            Node pre=now.pre;
            Node next=now.next;
            pre.next=next;
            next.pre=pre;
        }
    }

    public static void main(String[] args) {
        AllOne allOne=new AllOne();
        allOne.inc("hello");
        allOne.inc("hello");

        System.out.println(allOne.getMaxKey());
        System.out.println(allOne.getMinKey());
        allOne.inc("aaa");
        allOne.inc("aaa");
        allOne.dec("aaa");
        System.out.println(allOne.getMaxKey());
        System.out.println(allOne.getMinKey());
    }
}
